src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs

```diff
@@ -766,111 +760,184 @@ private void ProcessGoAwayFrame(FrameHeader frameHeader)
+                if (_abortException is not null)
                 {
-                    int activeBufferLength = _outgoingBuffer.ActiveLength;
-                    if (writeBytes >= totalBufferLength - activeBufferLength && activeBufferLength > 0)
-                    {
-                        // We explicitly do not pass cancellationToken here, as this flush impacts more than just this operation.
-                        await new ValueTask(FlushOutgoingBytesAsync()).ConfigureAwait(false); // await ValueTask to minimize number of awaiter fields
-                    }
+                    ThrowRequestAborted(_abortException);
                 }
 
-                // Invoke the callback with the supplied state and the target write buffer.
-                _outgoingBuffer.EnsureAvailableSpace(writeBytes);
-                FlushTiming flush = lockedAction(state, _outgoingBuffer.AvailableMemorySliced(writeBytes));
+                _writeQueue.Enqueue(writeEntry);
 
-                // Finish the write
-                _outgoingBuffer.Commit(writeBytes);
-                _lastPendingWriterShouldFlush |= flush == FlushTiming.AfterPendingWrites;
-                EndWrite(forceFlush: flush == FlushTiming.Now);
-            }
-            finally
-            {
-                _writerLock.Exit();
+                if (_writeSignal is not null)
+                {
+                    _writeSignal.SetResult();
+                    _writeSignal = null;
+                }
             }
+
+            return writeEntry.Task;
         }
 
-        private void EndWrite(bool forceFlush)
+        private async Task ProcessOutgoingFramesAsync()
         {
-            // We can't validate that we hold the mutex, but we can at least validate that someone is holding it.
-            Debug.Assert(_writerLock.IsHeld);
-
-            // We must flush if the caller requires it or if this or a recent frame wanted to be flushed
-            // once there were no more pending writers that themselves could have forced the flush.
-            if (forceFlush || (_pendingWriters == 0 && _lastPendingWriterShouldFlush))
+            try
             {
-                Debug.Assert(_inProgressWrite == null);
-                if (_outgoingBuffer.ActiveLength > 0)
+                while (true)
+                {
+                    WriteQueueEntry? writeEntry;
+                    Task? waitForSignal = null;
+
+                    lock (SyncObject)
+                    {
+                        Debug.Assert(_writeSignal is null);
+
+                        if (_abortException is not null)
+                        {
+                            // Connection is aborting, so stop processing
+                            break;
+                        }
+
+                        if (!_writeQueue.TryDequeue(out writeEntry))
+                        {
+                            _writeSignal = new TaskCompletionSource(TaskCreationOptions.RunContinuationsAsynchronously);
```

Comment:
Rather than manually managing this `Queue<T>` and a `_writeSignal`, could this instead use a `Channel<T>`?  That already handles being efficient about allocation.  And you may even be able to reduce synchronization if, for example, you're able to specify SingleReader=true when creating the unbounded channel (which from skimming this PR seems like it should be feasible, as there's just the single reader processing the queue).

Reply:
I can try that.